lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -490,25 +490,62 @@ LogicalResult LowerAnnotationsPass::solveWiringProblems(ApplyState &state) {
     return value.getDefiningOp()->getParentOfType<FModuleLike>();
   };
 
+  // Utility function to connect a destination to a source.  Always use a
+  // ConnectOp as the widths may be uninferred.
+  auto connect = [&](Value src, Value dest, ImplicitLocOpBuilder &builder) {
+    if (foldFlow(dest) == Flow::Source)
+      std::swap(src, dest);
+    // Create RefSend/RefResolve if necessary.
+    if (dest.getType().isa<RefType>() && !src.getType().isa<RefType>()) {
+      src = builder.create<RefSendOp>(src);
+    } else if (!dest.getType().isa<RefType>() && src.getType().isa<RefType>()) {
+      src = builder.create<RefResolveOp>(src);
+    }
+    // If the sink is a wire with no users, then convert this to a node.
+    auto destOp = dyn_cast_or_null<WireOp>(dest.getDefiningOp());
+    if (destOp && dest.getUses().empty()) {
+      builder.create<NodeOp>(src.getType(), src, destOp.getName())
+          .setAnnotationsAttr(destOp.getAnnotations());
+      destOp->erase();
+      return;
+    }
+    // Otherwise, just connect to the source.
+    builder.create<ConnectOp>(dest, src);
+  };
+
   auto &instanceGraph = state.instancePathCache.instanceGraph;
   auto *context = state.circuit.getContext();
 
   // Examine all discovered Wiring Problems to determine modifications that need
   // to be made per-module.
   LLVM_DEBUG({ llvm::dbgs() << "Analyzing wiring problems:\n"; });
   DenseMap<FModuleLike, ModuleModifications> moduleModifications;
+  DenseSet<Value> sinks(state.wiringProblems.size());
   for (auto [index, problem] : llvm::enumerate(state.wiringProblems)) {
     // This is a unique index that is assigned to this specific wiring problem
     // and is used as a key during wiring to know which Values (ports, sources,
     // or sinks) should be connected.
     auto source = problem.source;
     auto sink = problem.sink;
 
+    // Check that no WiringProblems are trying to use the same sink.  This
+    // should never happen.
+    auto [oldSink, added] = sinks.insert(sink);
+    if (!added) {
+      auto diag = mlir::emitError(sink.getLoc())
+                  << "This sink is involved with a Wiring Problem which is "
+                     "also used by another Wiring Problem (this is both "
+                     "illegal and should be impossible).";
+      diag.attachNote(source.getLoc()) << "The source is here.";
+      diag.attachNote(oldSink->getLoc()) << "The other source is here.";
```

Comment:
```suggestion

```

`oldSink` isn't "The other source", and possibly has been erased and so should not be accessed in the error message, I think?

Reply:
You are correct! I reworked this slightly to: (1) not print the old source (as that requires also tracking old sources) and (2) to delay deletion of sink `WireOp`s until the end (at the cost of maintaining a list of ops to delete.

I am really still _not_ happy with this error message and this would be totally unacceptable if this was exposed to work with Chisel's `BoringUtils` API (which I'm pretty sure will let you, or not stop you from, boring to the same sink). We'll need to clean this up in the future, but this should be good enough to catch known weird situations.